nginx: if root is in Nix store, use path's hash as ETag

lukateras opened the pull request:
###### Motivation for this change

Resolves #25485. Usage example:

```
$ realpath /var/www
/nix/store/wnrhnnpdj3x50j5xz38zp1qxs1ygwccw-site
$ curl --head localhost
HTTP/1.1 200 OK
Server: nginx
Date: Fri, 28 Sep 2018 06:09:25 GMT
Content-Type: text/html
Content-Length: 50
Last-Modified: Thu, 01 Jan 1970 00:00:01 GMT
Connection: keep-alive
ETag: "wnrhnnpdj3x50j5xz38zp1qxs1ygwccw"
Accept-Ranges: bytes
```

cc @Akii @chris-martin @domenkozar  

P.S. I didn't commit patch into Nixpkgs repo because that seems to be the preference, but it would be nice to have the patch reviewed. Tell me if I should commit it anyway.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

lukateras committed:
nginx: if root is in Nix store, use path's hash as ETag

Resolves #25485. Usage example:

$ realpath /var/www
/nix/store/wnrhnnpdj3x50j5xz38zp1qxs1ygwccw-site
$ curl --head localhost
HTTP/1.1 200 OK
Server: nginx
Date: Fri, 28 Sep 2018 06:09:25 GMT
Content-Type: text/html
Content-Length: 50
Last-Modified: Thu, 01 Jan 1970 00:00:01 GMT
Connection: keep-alive
ETag: "wnrhnnpdj3x50j5xz38zp1qxs1ygwccw"
Accept-Ranges: bytes

yegortimoshenko commented on pkgs/servers/http/nginx/generic.nix:
```diff
@@ -17,6 +17,12 @@ stdenv.mkDerivation {
   };
 
+  patches = [
+    (fetchpatch {
```
This could actually be `fetchurl`, but I'm not sure if I should use here or leave `fetchpatch` for consistency.

xeji replied:
I think it would be better to add this patch to nixpkgs repo. The general preference (as I understand it) is to use `fetchpatch` for patches from upstream, other distros or other authors, while patches written specifically for nixpkgs should be added to our repo.

yegortimoshenko replied:
Maybe. @grahamc, what do you think?

GrahamcOfBorg commented:
<!--REQUEST_ID=75feae72-d485-4e81-bf4e-94515ff1b8a9-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48337&attempt_id=f6a7f006-23d6-472f-855e-fcd32f361139)

Attempted: nginx

<details><summary>Partial log (click to expand)</summary><p>

```
test -d '/nix/store/r4bzjgc4y1k1f27p2g4a3660pydg6m11-nginx-1.14.0/logs' \
        || mkdir -p '/nix/store/r4bzjgc4y1k1f27p2g4a3660pydg6m11-nginx-1.14.0/logs'
make[1]: Leaving directory '/build/nginx-1.14.0'
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/r4bzjgc4y1k1f27p2g4a3660pydg6m11-nginx-1.14.0
shrinking /nix/store/r4bzjgc4y1k1f27p2g4a3660pydg6m11-nginx-1.14.0/bin/nginx
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/r4bzjgc4y1k1f27p2g4a3660pydg6m11-nginx-1.14.0/bin
patching script interpreter paths in /nix/store/r4bzjgc4y1k1f27p2g4a3660pydg6m11-nginx-1.14.0
checking for references to /build in /nix/store/r4bzjgc4y1k1f27p2g4a3660pydg6m11-nginx-1.14.0...
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=75feae72-d485-4e81-bf4e-94515ff1b8a9-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48337&attempt_id=2c5892c7-7557-4811-9af1-a1c0f8fb58db)

Attempted: nginx

<details><summary>Partial log (click to expand)</summary><p>

```
test -d '/nix/store/vkdh0nnyfg6j9fdmky7x2img0vgx0jzz-nginx-1.14.0/logs' \
        || mkdir -p '/nix/store/vkdh0nnyfg6j9fdmky7x2img0vgx0jzz-nginx-1.14.0/logs'
make[1]: Leaving directory '/build/nginx-1.14.0'
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/vkdh0nnyfg6j9fdmky7x2img0vgx0jzz-nginx-1.14.0
shrinking /nix/store/vkdh0nnyfg6j9fdmky7x2img0vgx0jzz-nginx-1.14.0/bin/nginx
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/vkdh0nnyfg6j9fdmky7x2img0vgx0jzz-nginx-1.14.0/bin
patching script interpreter paths in /nix/store/vkdh0nnyfg6j9fdmky7x2img0vgx0jzz-nginx-1.14.0
checking for references to /build in /nix/store/vkdh0nnyfg6j9fdmky7x2img0vgx0jzz-nginx-1.14.0...
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=75feae72-d485-4e81-bf4e-94515ff1b8a9-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48337&attempt_id=37ec6eb6-9757-406e-abb6-4f7c1fd37744)

Attempted: nginx

<details><summary>Partial log (click to expand)</summary><p>

```
        || mkdir -p '/nix/store/nc4hcpmi05q8wb0f4hp56q5cn72lip4d-nginx-1.14.0/logs'
test -d '/nix/store/nc4hcpmi05q8wb0f4hp56q5cn72lip4d-nginx-1.14.0/html' \
        || cp -R html '/nix/store/nc4hcpmi05q8wb0f4hp56q5cn72lip4d-nginx-1.14.0'
test -d '/nix/store/nc4hcpmi05q8wb0f4hp56q5cn72lip4d-nginx-1.14.0/logs' \
        || mkdir -p '/nix/store/nc4hcpmi05q8wb0f4hp56q5cn72lip4d-nginx-1.14.0/logs'
make[1]: Leaving directory '/private/tmp/nix-build-nginx-1.14.0.drv-0/nginx-1.14.0'
post-installation fixup
strip is /nix/store/53nysl8bqwxihwzs1pgwka20nf8mbvlp-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/nc4hcpmi05q8wb0f4hp56q5cn72lip4d-nginx-1.14.0/bin
patching script interpreter paths in /nix/store/nc4hcpmi05q8wb0f4hp56q5cn72lip4d-nginx-1.14.0
```
</p></details>

yegortimoshenko commented:
Yes, I believe so. Browsers will use ETag as a validation token to ensure content didn't change, and that maps onto Nix store semantics very well. There is also some syntax involved and that's conformant too (double quotes, optionally weak prefix but it doesn't apply here).

This patch seems to generate [RFC 7232](https://tools.ietf.org/html/rfc7232#section-2.3) compliant tags.

Additionally, Nginx will also properly handle this ETag and respond with `304 Not Modified` if client sends HEAD request with `If-None-Match: "<ETag>"` header.

yegortimoshenko commented:
Included patch into the PR.

domenkozar commented:
I think whatever we do here, it shouldn't be the default. We can make it convenient to turn on Nix+Nginx behaviour for those that understand what it does.

yegortimoshenko commented:
Why? If Nix store symlinks are handled properly, this shouldn't cause any regressions whatsoever.

gebner commented on pkgs/servers/http/nginx/nix-etag-1.15.4.patch:
```diff
@@ -0,0 +1,87 @@
++    #define NIX_STORE_DIR "/nix/store"
++    #define NIX_STORE_LEN ngx_strlen(NIX_STORE_DIR)
++
++    if (!ngx_strncmp(real, NIX_STORE_DIR, NIX_STORE_LEN)
```
Does this segfault if `ngx_realpath` fails and hence `real` is null?

yegortimoshenko replied:
It's undefined behavior. Thanks for the catch! It won't segfault on GCC, but will on Clang.

gebner replied:
Oh, it segfaults very well.  It just doesn't happen in your example since `strncmp` is a pure function.  The compiler then [optimizes it away](https://godbolt.org/z/oBa3bT) because the result is not used.
```c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define STRING "abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz"

int main() {
  // segfault
  printf("%i\n", strncmp(NULL, STRING, strlen(STRING)));
}
```

yegortimoshenko replied:
Right, I've edited post above :-) Also, fixed in https://raw.githubusercontent.com/NixOS/nixpkgs/971f8e2c779f57a3dfc931161d266b2c32b1d67e/pkgs/servers/http/nginx/nix-etag-1.15.4.patch.

gebner commented on pkgs/servers/http/nginx/nix-etag-1.15.4.patch:
```diff
@@ -0,0 +1,87 @@
++         ptr1 = real + NIX_STORE_LEN;
++        *ptr1 = '"';
++
++        ptr2 = (u_char *) ngx_strchr(ptr1, '-');
```
What if `real` is `/nix/store`?

yegortimoshenko replied:
In that case, it doesn't match. See `&& real[NIX_STORE_LEN] == '/'` in condition above.

gebner commented on pkgs/servers/http/nginx/nix-etag-1.15.4.patch:
```diff
@@ -0,0 +1,87 @@
+-        etag->hash = 0;
+-        return NGX_ERROR;
+-    }
++    real = ngx_realpath(clcf->root.data, NULL);
```
If `clcf->root.data` starts with `/nix/store/`, does this actually mean that the current request is a static file inside that root?  Can you break it using the `alias` directive?  (From my cursory reading of the nginx source code, `proxy_pass` seems to be fine since it bypasses this method entirely.)
```nginx
root /nix/store/...;

location /foo {
  alias /srv/foo;
}
```

yegortimoshenko replied:
It works as expected! Here's how to reproduce:

```
$ cat /etc/nixos/configuration.nix
{ pkgs, ... }:

let
  srv = pkgs.runCommand "srv" {} ''
    mkdir $out && echo "<!DOCTYPE html><b>Hello :-)</b>" > $out/index.html
  '';
in

{
  # ...

  services.nginx = {
    enable = true;
    package = pkgs.nginx.overrideAttrs (self: {
      patches = (self.patches or []) ++ [ ./nix-etag-1.15.4.patch ];
    });
    virtualHosts.default = {
      root = srv;
      locations."/foo".alias = "/tmp/srv";
      listen = [({ addr = "127.0.0.1"; port = 8080; })];
    };
  };
}
$ cd /etc/nixos; sudo curl -LO https://raw.githubusercontent.com/NixOS/nixpkgs/971f8e2c779f57a3dfc931161d266b2c32b1d67e/pkgs/servers/http/nginx/nix-etag-1.15.4.patch
$ mkdir -p /tmp/srv
$ echo "<!DOCTYPE html><i>correct</i>" > /tmp/srv/index.html
$ sudo nixos-rebuild switch
$ curl --head http://localhost:8080
HTTP/1.1 200 OK
Server: nginx
Date: Tue, 16 Oct 2018 12:46:29 GMT
Content-Type: text/html
Content-Length: 32
Last-Modified: Thu, 01 Jan 1970 00:00:01 GMT
Connection: keep-alive
ETag: "rxawlg1ivq2zjz9qgfz9anvrwwcj4j9q"
Accept-Ranges: bytes
$ curl --head http://localhost:8080/foo/
HTTP/1.1 200 OK
Server: nginx
Date: Tue, 16 Oct 2018 12:46:41 GMT
Content-Type: text/html
Content-Length: 15
Last-Modified: Tue, 16 Oct 2018 12:43:00 GMT
Connection: keep-alive
ETag: "5bc5dcd4-f"
Accept-Ranges: bytes
```

Meaning, `alias` requests override `root`.

lukateras committed:
nginx: check for realpath() == NULL in ETag patch

Thanks to Gabriel Ebner!

GrahamcOfBorg commented:
<!--REQUEST_ID=cdc63037-1a15-4c49-84ae-cc3984728b6c-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48337&attempt_id=3be14422-d73a-4a2c-bfdc-10f2f6b2b483)

Attempted: nginx

<details><summary>Partial log (click to expand)</summary><p>

```
test -d '/nix/store/3hlqw53imrs5l7v34gj4i6f0xi2sjq09-nginx-1.14.0/logs' \
        || mkdir -p '/nix/store/3hlqw53imrs5l7v34gj4i6f0xi2sjq09-nginx-1.14.0/logs'
make[1]: Leaving directory '/build/nginx-1.14.0'
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/3hlqw53imrs5l7v34gj4i6f0xi2sjq09-nginx-1.14.0
shrinking /nix/store/3hlqw53imrs5l7v34gj4i6f0xi2sjq09-nginx-1.14.0/bin/nginx
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/3hlqw53imrs5l7v34gj4i6f0xi2sjq09-nginx-1.14.0/bin
patching script interpreter paths in /nix/store/3hlqw53imrs5l7v34gj4i6f0xi2sjq09-nginx-1.14.0
checking for references to /build in /nix/store/3hlqw53imrs5l7v34gj4i6f0xi2sjq09-nginx-1.14.0...
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=cdc63037-1a15-4c49-84ae-cc3984728b6c-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48337&attempt_id=b5be0864-94fe-4f48-902f-c3e4b939513f)

Attempted: nginx

<details><summary>Partial log (click to expand)</summary><p>

```
test -d '/nix/store/gcgzsqbr0shcp55p7lcp02wm24hwwg07-nginx-1.14.0/logs' \
        || mkdir -p '/nix/store/gcgzsqbr0shcp55p7lcp02wm24hwwg07-nginx-1.14.0/logs'
make[1]: Leaving directory '/build/nginx-1.14.0'
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/gcgzsqbr0shcp55p7lcp02wm24hwwg07-nginx-1.14.0
shrinking /nix/store/gcgzsqbr0shcp55p7lcp02wm24hwwg07-nginx-1.14.0/bin/nginx
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/gcgzsqbr0shcp55p7lcp02wm24hwwg07-nginx-1.14.0/bin
patching script interpreter paths in /nix/store/gcgzsqbr0shcp55p7lcp02wm24hwwg07-nginx-1.14.0
checking for references to /build in /nix/store/gcgzsqbr0shcp55p7lcp02wm24hwwg07-nginx-1.14.0...
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=cdc63037-1a15-4c49-84ae-cc3984728b6c-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48337&attempt_id=3ceeef74-b56d-47c9-b5e0-bac47ca5b0f0)

Attempted: nginx

<details><summary>Partial log (click to expand)</summary><p>

```
        || mkdir -p '/nix/store/pys7fcv41sbyj3qqr27lmab3k9p2cyvk-nginx-1.14.0/logs'
test -d '/nix/store/pys7fcv41sbyj3qqr27lmab3k9p2cyvk-nginx-1.14.0/html' \
        || cp -R html '/nix/store/pys7fcv41sbyj3qqr27lmab3k9p2cyvk-nginx-1.14.0'
test -d '/nix/store/pys7fcv41sbyj3qqr27lmab3k9p2cyvk-nginx-1.14.0/logs' \
        || mkdir -p '/nix/store/pys7fcv41sbyj3qqr27lmab3k9p2cyvk-nginx-1.14.0/logs'
make[1]: Leaving directory '/private/tmp/nix-build-nginx-1.14.0.drv-0/nginx-1.14.0'
post-installation fixup
strip is /nix/store/53nysl8bqwxihwzs1pgwka20nf8mbvlp-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/pys7fcv41sbyj3qqr27lmab3k9p2cyvk-nginx-1.14.0/bin
patching script interpreter paths in /nix/store/pys7fcv41sbyj3qqr27lmab3k9p2cyvk-nginx-1.14.0
```
</p></details>

lukateras committed:
nginx: handle impure symlinks in ETag patch

GrahamcOfBorg commented:
<!--REQUEST_ID=97006bf0-c738-4bce-9e64-46992b90535e-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48337&attempt_id=e62ad896-2fd8-41b3-a496-34f31a6e5c84)

Attempted: nginx

<details><summary>Partial log (click to expand)</summary><p>

```
test -d '/nix/store/d1mpq32i9qyvpiw1x4qq3dj2k3w6r0w6-nginx-1.14.0/logs' \
        || mkdir -p '/nix/store/d1mpq32i9qyvpiw1x4qq3dj2k3w6r0w6-nginx-1.14.0/logs'
make[1]: Leaving directory '/build/nginx-1.14.0'
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/d1mpq32i9qyvpiw1x4qq3dj2k3w6r0w6-nginx-1.14.0
shrinking /nix/store/d1mpq32i9qyvpiw1x4qq3dj2k3w6r0w6-nginx-1.14.0/bin/nginx
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/d1mpq32i9qyvpiw1x4qq3dj2k3w6r0w6-nginx-1.14.0/bin
patching script interpreter paths in /nix/store/d1mpq32i9qyvpiw1x4qq3dj2k3w6r0w6-nginx-1.14.0
checking for references to /build in /nix/store/d1mpq32i9qyvpiw1x4qq3dj2k3w6r0w6-nginx-1.14.0...
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=97006bf0-c738-4bce-9e64-46992b90535e-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48337&attempt_id=dec9bfbb-37f1-4337-a6cb-6290b0dd7b17)

Attempted: nginx

<details><summary>Partial log (click to expand)</summary><p>

```
test -d '/nix/store/7nsm2sz5q5xilp0kqhzqa33a99n2nnr7-nginx-1.14.0/logs' \
        || mkdir -p '/nix/store/7nsm2sz5q5xilp0kqhzqa33a99n2nnr7-nginx-1.14.0/logs'
make[1]: Leaving directory '/build/nginx-1.14.0'
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/7nsm2sz5q5xilp0kqhzqa33a99n2nnr7-nginx-1.14.0
shrinking /nix/store/7nsm2sz5q5xilp0kqhzqa33a99n2nnr7-nginx-1.14.0/bin/nginx
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/7nsm2sz5q5xilp0kqhzqa33a99n2nnr7-nginx-1.14.0/bin
patching script interpreter paths in /nix/store/7nsm2sz5q5xilp0kqhzqa33a99n2nnr7-nginx-1.14.0
checking for references to /build in /nix/store/7nsm2sz5q5xilp0kqhzqa33a99n2nnr7-nginx-1.14.0...
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=97006bf0-c738-4bce-9e64-46992b90535e-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48337&attempt_id=ecb0efbe-1154-46cf-af9c-09c6ae105174)

Attempted: nginx

<details><summary>Partial log (click to expand)</summary><p>

```
        || mkdir -p '/nix/store/cpq2ca2652p6icr2xxprlk49qqgnqj8d-nginx-1.14.0/logs'
test -d '/nix/store/cpq2ca2652p6icr2xxprlk49qqgnqj8d-nginx-1.14.0/html' \
        || cp -R html '/nix/store/cpq2ca2652p6icr2xxprlk49qqgnqj8d-nginx-1.14.0'
test -d '/nix/store/cpq2ca2652p6icr2xxprlk49qqgnqj8d-nginx-1.14.0/logs' \
        || mkdir -p '/nix/store/cpq2ca2652p6icr2xxprlk49qqgnqj8d-nginx-1.14.0/logs'
make[1]: Leaving directory '/private/tmp/nix-build-nginx-1.14.0.drv-0/nginx-1.14.0'
post-installation fixup
strip is /nix/store/53nysl8bqwxihwzs1pgwka20nf8mbvlp-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/cpq2ca2652p6icr2xxprlk49qqgnqj8d-nginx-1.14.0/bin
patching script interpreter paths in /nix/store/cpq2ca2652p6icr2xxprlk49qqgnqj8d-nginx-1.14.0
```
</p></details>

domenkozar commented:
@yegortimoshenko because we're changing expected behavior in nginx, it could have implications we can't predict. Although to be fair, current "cache forever" is pretty bad so maybe it's not that bad :)

grahamc commented:
We should probably merge this.

aszlig commented on pkgs/servers/http/nginx/nix-etag-1.15.4.patch:
```diff
@@ -0,0 +1,90 @@
+-                                  r->headers_out.content_length_n)
+-                      - etag->value.data;
++    #define NIX_STORE_DIR "/nix/store"
++    #define NIX_STORE_LEN ngx_strlen(NIX_STORE_DIR)
```
Maybe this could be determined during configure phase along with the length by looking up the `NIX_STORE` environment variable.

yegortimoshenko replied:
Could this be `postPatch`:
```
substituteInPlace \
  --replace 'NIX_STORE_DIR "/nix/store"' "NIX_STORE_DIR \"$NIX_STORE\""
```

aszlig replied:
Sure, I just thought it would be better if we'd have it directly in `configure`, so that if people want to disable this all they'd need to do is remove the patch. And as said, I would probably also add the NIX_STORE_LEN so that we save a `ngx_strlen` at runtime.

aszlig replied:
Maybe even use `@NIX_STORE_...@` in the patch and substitute with:

```
substituteInPlace ... \
  --subst-var-by NIX_STORE_DIR "$NIX_STORE" \
  --subst-var-by NIX_STORE_LEN "$#NIX_STORE"
```

aszlig commented on pkgs/servers/http/nginx/nix-etag-1.15.4.patch:
```diff
@@ -0,0 +1,90 @@
++                                      r->headers_out.content_length_n)
++                          - etag->value.data;
++    }
+ 
```
Probably add `ngx_free(real)` here, so we don't introduce a memleak.

aszlig commented on pkgs/servers/http/nginx/nix-etag-1.15.4.patch:
```diff
@@ -0,0 +1,90 @@
++        && real[NIX_STORE_LEN] == '/'
++        && real[NIX_STORE_LEN + 1] != '\0')
++    {
++         ptr1 = real + NIX_STORE_LEN;
```
Nitpick: Indentation.

pbogdan commented:
Hmm, could they (assuming they are conforming):

> - If an entity tag has been provided by the origin server, **MUST**
>        use that entity tag in any cache-conditional request (using If-
>        Match or If-None-Match).

domenkozar commented:
@GrahamcOfBorg build nginx

domenkozar commented:
Ah, bad merge. @aszlig does this look good to go?

aszlig commented:
@domenkozar: Well, more or less, the memory leak would be great to fix.

aszlig committed:
nginx: Fix memleak in nix-etag patch

The original patch introduced a new "real" variable which gets populated
(and allocated) via ngx_realpath(). It's properly freed in error
conditions but it won't be freed if ngx_http_set_etag returns
successfully.

Adding another ngx_free() just before returning fixes that memory leak.

I also fixed a small indentation issue along the way.

Signed-off-by: aszlig <aszlig@nix.build>

aszlig commented:
@domenkozar: Addressed the mentioned issue, but I'll also add a test for this, so please don't merge yet :-)

aszlig committed:
nixos/tests/nginx: Add subtest for Nix ETag patch

This is to make sure that we get different ETag values whenever we
switch to a different store path but with the same file contents.

I've checked this against the old behaviour without the patch and it
fails as expected.

Signed-off-by: aszlig <aszlig@nix.build>

aszlig commented:
@domenkozar: Added a subtest to the `nginx` NixOS VM test and rebased against `master`, because of a conflict in the test expression.

aszlig commented:
@grahamcofborg test nginx

aszlig committed:
nginx/etag-patch: Use Nix store dir from build env

So far, the Nix store directory was hardcoded and if someone uses a
different Nix store directory the patch won't work. Of course, this is
pretty uncommon, but by not only substituting the store directory but
also the length of it we also save a few calls to ngx_strlen(), which
should save us a few cycles.

Signed-off-by: aszlig <aszlig@nix.build>

aszlig commented:
@domenkozar: Instead of hardcoding the Nix store directory, it's now looked up from the builder environment.

@yegortimoshenko: I've hijacked your patch by only changing the hunks necessary for the changes, please ack if you're okay with that.

@GrahamcOfBorg test nginx

So from my point of view, this should be ready to be merged after the ack from Yegor :smiley:

yegortimoshenko commented:
@aszlig Ack! Thank you for working on this.

aszlig commented:
Heads-up everyone: After testing this in a few production instances, it seems that some browsers still get cache hits for new store paths (and changed contents) for some reason. I highly suspect that it might be due to the `last-modified` header (as mentioned in https://github.com/NixOS/nixpkgs/pull/48337#issuecomment-451644084).

Going to test this with `last-modified` disabled for a little while and if this is the case I think we should improve that patch by disabling `last-modified` if serving from a store path.

yegortimoshenko commented:
ETag should take precedence over Last-Modified: https://stackoverflow.com/questions/824152/what-takes-precedence-the-etag-or-last-modified-http-header/1560098

aszlig commented:
@yegortimoshenko: Yep, in theory it should, see also https://github.com/NixOS/nixpkgs/pull/48337#issuecomment-451646135.

juskrey commented:
Some browsers indeed show very strange caching behavior when both ETag and last-modified are present. At least Chrome seems to keep etag-ged file for a while when last-modified is old

aszlig commented:
@juskrey: Mhm, I just could confirm exactly this. I think we also need to disable last-modified if we provide an ETag from the store.

aszlig commented:
[Here](https://gist.github.com/aszlig/83dfe23f96bcae2116d5a0a90e46ebd2) is a small NixOS VM test reproducing the issue, pull request with the test and fix following soon...

aszlig commented:
Done: https://github.com/NixOS/nixpkgs/pull/76697
